distributed/s3.py

```diff
@@ -26,148 +32,635 @@
     -------
-    list of Futures.  Each future holds bytes for one key within the bucket
-
-    Examples
-    --------
-    >>> futures = read_bytes('distributed-test', 'test')  # doctest: +SKIP
-    >>> futures  # doctest: +SKIP
-    [<Future: status: finished, key: read_content_from_keys-00092e8a75141837c1e9b717b289f9d2>,
-     <Future: status: finished, key: read_content_from_keys-4f0f2cbcf4573a373cc62467ffbfd30d>]
-    >>> futures[0].result()  # doctest: +SKIP
-    b'{"amount": 100, "name": "Alice"}\\n{"amount": 200, "name": "Bob"}\\n
-      {"amount": 300, "name": "Charlie"}\\n{"amount": 400, "name": "Dennis"}\\n'
+    List of ``distributed.Future`` objects if ``lazy=False``
+    or ``dask.Value`` objects if ``lazy=True``
     """
+    if s3 is None:
+        s3 = S3FileSystem(**s3pars)
     executor = default_executor(executor)
-    s3_objects = get_list_of_summary_objects(bucket_name, prefix,
-                                             path_delimiter, anon=anon)
-    keys = [obj.key for obj in s3_objects]
-
-    names = ['read-bytes-{0}'.format(key) for key in keys]
-
+    filenames, lengths, offsets = [], [], []
+    for afile in s3.glob(fn):
+        size = s3.info(afile)['Size']
+        offset = list(range(0, size, blocksize))
+        if not_zero:
+            offset[0] = 1
+        offsets.extend(offset)
+        filenames.extend([afile]*len(offset))
+        lengths.extend([blocksize]*len(offset))
+    names = ['read-binary-s3-%s-%s' % (fn, tokenize(offset, length, delimiter, not_zero))
+            for fn, offset, length in zip(filenames, offsets, lengths)]
+
+    logger.debug("Read %d blocks of binary bytes from %s", len(names), fn)
     if lazy:
-        values = [Value(name, [{name: (read_content_from_keys, bucket_name,
-                                       key, anon)}])
-                  for name, key in zip(names, keys)]
+        executor._send_to_scheduler({'op': 'update-graph',
+                                     'tasks': {},
+                                     'dependencies': set(),
+                                     'keys': [],
+                                     'client': executor.id})
+        values = [Value(name, [{name: (read_block_from_s3, fn, offset, length, s3pars, delimiter)}])
+                  for name, fn, offset, length in zip(names, filenames, offsets, lengths)]
         return values
     else:
-        return executor.map(read_content_from_keys, [bucket_name] * len(keys),
-                keys, anon=anon)
+        return executor.map(read_block_from_s3, filenames, offsets, lengths,
+                s3pars=s3pars, delimiter=delimiter)
 
 
-def read_text(bucket_name, prefix='', path_delimiter='', encoding='utf-8',
-        errors='strict', lineterminator='\n', executor=None, anon=None,
-        collection=True, lazy=True, compression=None):
-    """
-    Read lines of text from S3
+def read_block_from_s3(filename, offset, length, s3pars={}, delimiter=None):
+    s3 = S3FileSystem(**s3pars)
```

Comment:
The S3FileSystem does cache connections and directory entries; but actually boto3 establishes new HTTP connections for everything it does.